extensions/eclipse-che-theia-git-provisioner/src/node/git-configuration-controller.ts

```diff
@@ -0,0 +1,128 @@
+import { Disposable } from '@theia/core';
+
+export const GIT_CONFIG_PATH = resolve(homedir(), '.gitconfig');
+
+export interface UserConfiguration {
+    name: string | undefined;
+    email: string | undefined;
+}
+
+@injectable()
+export class GitConfigurationController {
+
+    @inject(CheTheiaUserPreferencesSynchronizer)
+    protected preferencesService: CheTheiaUserPreferencesSynchronizer;
+
+    protected preferencesHandler: Disposable | undefined;
+
+    protected gitConfigWatcher: nsfw.NSFW | undefined;
+
+    public async watchGitConfigChanges(): Promise<void> {
+        if (this.gitConfigWatcher) {
+            return;
+        }
+
+        const gitConfigExists = await pathExists(GIT_CONFIG_PATH);
+        if (!gitConfigExists) {
+            await createFile(GIT_CONFIG_PATH);
+        }
+
+        this.gitConfigWatcher = await nsfw(GIT_CONFIG_PATH, async (events: nsfw.ChangeEvent[]) => {
+            for (const event of events) {
+                if (event.action === nsfw.actions.MODIFIED) {
+                    const userConfig = await this.getUserConfigurationFromGitConfig();
+                    const preferences = await this.preferencesService.getPreferences();
+
+                    preferences['git.user.name'] = userConfig.name;
+                    preferences['git.user.email'] = userConfig.email;
+
+                    await this.preferencesService.setPreferences(preferences);
+                }
+            }
+        });
+        await this.gitConfigWatcher.start();
+    }
+
+    protected async getUserConfigurationFromGitConfig(): Promise<UserConfiguration> {
+        const gitConfigExists = await pathExists(GIT_CONFIG_PATH);
+        if (!gitConfigExists) {
+            return {} as UserConfiguration;
+        }
+
+        const gitConfigContent = await readFile(GIT_CONFIG_PATH, 'utf-8');
+        const gitConfig = ini.parse(gitConfigContent);
+
+        if (gitConfig.user === undefined) {
+            return {} as UserConfiguration;
+        }
+
+        const userName = gitConfig.user.name;
+        const userEmail = gitConfig.user.email;
+
+        return { name: userName, email: userEmail };
+    }
+
+    public async watchUserPreferencesChanges(): Promise<void> {
+        if (this.preferencesHandler) {
+            return;
+        }
+
+        this.preferencesHandler = this.preferencesService.onUserPreferencesModify(preferences => {
+            const config = this.getUserConfiguration(preferences);
+            this.updateGlobalGitConfig(config);
+        });
+    }
+
+    protected getUserConfiguration(preferences: object): UserConfiguration {
+        return {
+            name: preferences['git.user.name'],
+            email: preferences['git.user.email']
+        };
+    }
+
+    protected async updateGlobalGitConfig(config: UserConfiguration): Promise<void> {
+        if (config.name === undefined && config.email === undefined) {
+            return;
+        }
+
+        const gitConfig = { user: {} };
```

Comment:
You can improve typization here:

```typescript
const gitConfig =  { user : {} as UserConfiguration };
 if (config.name) {
     gitConfig.user.name = config.name;
 }
...
```

Or create object and use it.:
```typescript
export interface GitConfig {
   user: UserConfiguration
}
```

Reply:
Thanks, updated.